File: backend/app/core/sessions.py
import hashlib
import json
import secrets
from datetime import datetime, timezone
from typing import Any
from uuid import UUID

from fastapi import Response
from starlette.requests import Request

from app.core.config import Settings, get_settings
from app.core.rate_limit import client_ip
from app.core.security import cookie_should_be_secure
from app.db import data_api
from app.db.dragonfly import get_dragonfly
from app.models import User

USER_SESSIONS_TTL = 60 * 60 * 24 * 90
TOUCH_EVERY_SECONDS = 300


def _session_key(token: str) -> str:
    return f"session:{token}"


def _user_sessions_key(user_id: str) -> str:
    return f"user_sessions:{user_id}"


def session_id_for(token: str) -> str:
    return hashlib.sha256(token.encode()).hexdigest()


def _now() -> str:
    return datetime.now(timezone.utc).isoformat()


async def create_session(user_id: str, remember: bool = False, request: Request | None = None) -> tuple[str, int]:
    settings = get_settings()
    token = secrets.token_urlsafe(32)
    ttl = settings.session_remember_ttl_seconds if remember else settings.session_ttl_seconds
    payload = {
        "user_id": user_id,
        "remember": remember,
        "created_at": _now(),
        "last_seen_at": _now(),
        "user_agent": ((request.headers.get("user-agent") or "")[:240] if request else ""),
        "ip": client_ip(request) if request else "",
    }
    redis = get_dragonfly()
    await redis.set(_session_key(token), json.dumps(payload), ex=ttl)
    await redis.sadd(_user_sessions_key(user_id), token)
    await redis.expire(_user_sessions_key(user_id), max(ttl, USER_SESSIONS_TTL))
    return token, ttl


async def destroy_session(token: str | None) -> None:
    if not token:
        return
    redis = get_dragonfly()
    data = await load_session(token)
    await redis.delete(_session_key(token))
    if data and data.get("user_id"):
        await redis.srem(_user_sessions_key(str(data["user_id"])), token)


async def load_session(token: str | None) -> dict[str, Any] | None:
    if not token:
        return None
    raw = await get_dragonfly().get(_session_key(token))
    if not raw:
        return None
    try:
        data = json.loads(raw)
    except json.JSONDecodeError:
        return None
    return data if isinstance(data, dict) and data.get("user_id") else None


async def touch_session(token: str, remember: bool) -> None:
    settings = get_settings()
    ttl = settings.session_remember_ttl_seconds if remember else settings.session_ttl_seconds
    redis = get_dragonfly()
    data = await load_session(token)
    if not data:
        return
    last = str(data.get("last_seen_at") or "")
    should_write = True
    if last:
        try:
            previous = datetime.fromisoformat(last)
            should_write = (datetime.now(timezone.utc) - previous).total_seconds() >= TOUCH_EVERY_SECONDS
        except ValueError:
            should_write = True
    if should_write:
        data["last_seen_at"] = _now()
        await redis.set(_session_key(token), json.dumps(data), ex=ttl)
    else:
        await redis.expire(_session_key(token), ttl)
    if data.get("user_id"):
        await redis.sadd(_user_sessions_key(str(data["user_id"])), token)
        await redis.expire(_user_sessions_key(str(data["user_id"])), USER_SESSIONS_TTL)


def attach_session_cookie(response: Response, request: Request, token: str, ttl: int, settings: Settings) -> None:
    response.set_cookie(
        key=settings.session_cookie_name,
        value=token,
        max_age=ttl,
        path="/",
        httponly=True,
        secure=cookie_should_be_secure(request, settings),
        samesite="lax",
    )


def clear_session_cookie(response: Response, request: Request, settings: Settings) -> None:
    response.delete_cookie(
        key=settings.session_cookie_name,
        path="/",
        httponly=True,
        secure=cookie_should_be_secure(request, settings),
        samesite="lax",
    )


async def list_sessions(user_id: str, current_token: str | None = None) -> list[dict[str, Any]]:
    redis = get_dragonfly()
    tokens = await redis.smembers(_user_sessions_key(user_id))
    sessions: list[dict[str, Any]] = []
    for token in tokens or []:
        data = await load_session(token)
        if not data:
            await redis.srem(_user_sessions_key(user_id), token)
            continue
        sessions.append({
            "id": session_id_for(token),
            "created_at": data.get("created_at"),
            "last_seen_at": data.get("last_seen_at") or data.get("created_at"),
            "user_agent": data.get("user_agent") or "",
            "ip": data.get("ip") or "",
            "current": bool(current_token) and len(token) == len(current_token) and secrets.compare_digest(token, current_token),
        })
    sessions.sort(key=lambda item: (item["current"], str(item.get("last_seen_at") or "")), reverse=True)
    return sessions


async def revoke_session(user_id: str, session_id: str) -> bool:
    redis = get_dragonfly()
    for token in await redis.smembers(_user_sessions_key(user_id)) or []:
        digest = session_id_for(token)
        if len(digest) == len(session_id) and secrets.compare_digest(digest, session_id):
            await destroy_session(token)
            return True
    return False


async def revoke_other_sessions(user_id: str, keep_token: str | None) -> int:
    redis = get_dragonfly()
    removed = 0
    for token in await redis.smembers(_user_sessions_key(user_id)) or []:
        if keep_token and len(token) == len(keep_token) and secrets.compare_digest(token, keep_token):
            continue
        await destroy_session(token)
        removed += 1
    return removed


async def revoke_all_sessions(user_id: str) -> int:
    return await revoke_other_sessions(user_id, None)


async def get_user_from_request(request: Request) -> User | None:
    settings = get_settings()
    token = request.cookies.get(settings.session_cookie_name)
    data = await load_session(token)
    if not data or not token:
        return None
    try:
        UUID(str(data["user_id"]))
    except ValueError:
        return None
    user = await data_api.get_user(str(data["user_id"]))
    if user is None:
        await destroy_session(token)
        return None
    if user.currently_suspended:
        await destroy_session(token)
        return None
    from app.db import admin_db
    if await admin_db.user_is_banned(user.id):
        await destroy_session(token)
        return None
    await touch_session(token, bool(data.get("remember")))
    return user

File: frontend/lib/server/sessions.ts
import "server-only";

import { createHash, randomBytes, timingSafeEqual } from "node:crypto";
import type { NextRequest, NextResponse } from "next/server";
import { userIsBanned } from "./account-bans";
import { cookieSecure } from "./cookies";
import { redis } from "./redis";
import { isSuspended, userById, type User } from "./users";

export const SESSION_COOKIE = process.env.MISA_SESSION_COOKIE_NAME || "misa_session";
const normalTtl = Number(process.env.MISA_SESSION_TTL_SECONDS || 604800);
const rememberedTtl = Number(process.env.MISA_SESSION_REMEMBER_TTL_SECONDS || 2592000);
const touchEveryMs = 5 * 60 * 1000;

type StoredSession = { user_id: string; remember?: boolean; created_at?: string; last_seen_at?: string; user_agent?: string; ip?: string };

const sessionKey = (token: string) => `session:${token}`;
const userSessionsKey = (userId: string) => `user_sessions:${userId}`;
const ttlFor = (remember: boolean) => remember ? rememberedTtl : normalTtl;

async function ready() {
  const client = redis();
  if (client.status === "wait") await client.connect();
  return client;
}

export async function createSession(userId: string, request: NextRequest, remember = false) {
  const token = randomBytes(32).toString("base64url");
  const now = new Date().toISOString();
  const payload: StoredSession = {
    user_id: userId,
    remember,
    created_at: now,
    last_seen_at: now,
    user_agent: (request.headers.get("user-agent") || "").slice(0, 240),
    ip: (request.headers.get("cf-connecting-ip") || request.headers.get("x-forwarded-for") || "").split(",")[0].trim(),
  };
  const client = await ready();
  await client.set(sessionKey(token), JSON.stringify(payload), "EX", ttlFor(remember));
  await client.sadd(userSessionsKey(userId), token);
  await client.expire(userSessionsKey(userId), Math.max(ttlFor(remember), 60 * 60 * 24 * 90));
  return { token, ttl: ttlFor(remember) };
}

export function attachSession(response: NextResponse, request: NextRequest, token: string, ttl: number) {
  response.cookies.set(SESSION_COOKIE, token, {
    maxAge: ttl,
    path: "/",
    httpOnly: true,
    sameSite: "lax",
    secure: cookieSecure(request),
  });
}

export function clearSession(response: NextResponse, request?: NextRequest) {
  response.cookies.set(SESSION_COOKIE, "", { maxAge: 0, path: "/", secure: request ? cookieSecure(request) : true, httpOnly: true, sameSite: "lax" });
}

async function sessionFor(token?: string | null) {
  if (!token) return null;
  const value = await (await ready()).get(sessionKey(token));
  if (!value) return null;
  try {
    const parsed = JSON.parse(value) as StoredSession;
    return parsed.user_id ? parsed : null;
  } catch { return null; }
}

export async function currentUser(request: NextRequest): Promise<User | null> {
  const token = request.cookies.get(SESSION_COOKIE)?.value;
  const session = await sessionFor(token);
  if (!session || !token) return null;
  const user = await userById(session.user_id);
  if (!user || isSuspended(user) || await userIsBanned(user)) {
    await destroySession(token, session.user_id);
    return null;
  }
  const lastSeen = Date.parse(session.last_seen_at || "");
  const client = await ready();
  if (Number.isNaN(lastSeen) || Date.now() - lastSeen >= touchEveryMs) {
    session.last_seen_at = new Date().toISOString();
    await client.set(sessionKey(token), JSON.stringify(session), "EX", ttlFor(Boolean(session.remember)));
  } else {
    await client.expire(sessionKey(token), ttlFor(Boolean(session.remember)));
  }
  await client.sadd(userSessionsKey(user.id), token);
  await client.expire(userSessionsKey(user.id), 7776000);
  return user;
}

export async function destroySession(token?: string | null, knownUserId?: string) {
  if (!token) return;
  const session = knownUserId ? null : await sessionFor(token);
  const userId = knownUserId || session?.user_id;
  const client = await ready();
  await client.del(sessionKey(token));
  if (userId) await client.srem(userSessionsKey(userId), token);
}

export function sessionId(token: string) {
  return createHash("sha256").update(token).digest("hex");
}

export function sameToken(left: string, right: string) {
  const a = Buffer.from(left);
  const b = Buffer.from(right);
  return a.length === b.length && timingSafeEqual(a, b);
}

export async function listSessions(userId: string, currentToken?: string | null) {
  const client = await ready(); const tokens = await client.smembers(userSessionsKey(userId)); const sessions: Array<{ id: string; created_at?: string; last_seen_at?: string; user_agent: string; ip: string; current: boolean }> = [];
  for (const token of tokens) { const session = await sessionFor(token); if (!session) { await client.srem(userSessionsKey(userId), token); continue; } sessions.push({ id: sessionId(token), created_at: session.created_at, last_seen_at: session.last_seen_at || session.created_at, user_agent: session.user_agent || "", ip: session.ip || "", current: Boolean(currentToken && sameToken(token, currentToken)) }); }
  return sessions.sort((a,b) => Number(b.current) - Number(a.current) || String(b.last_seen_at || "").localeCompare(String(a.last_seen_at || "")));
}

export async function revokeSessionById(userId: string, wanted: string) {
  const client = await ready(); const tokens = await client.smembers(userSessionsKey(userId));
  for (const token of tokens) if (sameToken(sessionId(token), wanted)) { await destroySession(token, userId); return true; }
  return false;
}

export async function revokeOtherSessions(userId: string, keep?: string | null) {
  const client = await ready(); const tokens = await client.smembers(userSessionsKey(userId)); let removed = 0;
  for (const token of tokens) { if (keep && sameToken(token, keep)) continue; await destroySession(token, userId); removed += 1; }
  return removed;
}
